### ISSUE TITLE
Support Graphing of Long (a.k.a. Tall) and Wide Format Time Series Dataframes

### ISSUE THREAD

**Issue by kylebrandt** [grafana/grafana#22219]
**What would you like to be added**:
Dataframes of certain formats will be graphed as time series. The following data can be graphed as is (screenshot uses react table visualization):

![image](https://user-images.githubusercontent.com/1692624/74556761-bdd17e80-4f2c-11ea-8b4a-437fbed8a2d9.png)

## Use Case

Currently our SQL plugins derive things from the special names of the columns. Instead we can use more conventional Long and Wide formats and derive the format from the schema. This will result in a more conventional SQL experience for users. The amount of code plugin authors need to create will be reduced.1

Taking a MySQL Example:

```sql
SELECT
  $__timeGroup(createdAt,'$summarize') as time_sec,
  avg(value) as value,
  hostname as metric
FROM 
  grafana_metric
WHERE
  $__timeFilter(createdAt) AND
  measurement = 'logins.count' AND
  hostname IN($host)
GROUP BY 1, 3
ORDER BY 1
```

Which becomes:

```sql
SELECT
  UNIX_TIMESTAMP(createdAt) DIV 60 * 60 as time_sec,
  avg(value) as value,
  hostname as metric
FROM 
  grafana_metric
WHERE
  createdAt BETWEEN FROM_UNIXTIME(1580908377) AND FROM_UNIXTIME(1580985308) AND
  measurement = 'logins.count' AND
  hostname IN('10.1.100.1','10.1.100.10','server1','server2')
GROUP BY 1, 3
ORDER BY 1
```

Produces Long Format Data:

```csv
"Time";"Metric";"Value"
"1580985300000";"server2";"445.35955810546875"
"1580985300000";"server1";"445.85382080078125"
"1580985300000";"10.1.100.10";"445.64312744140625"
"1580985300000";"10.1.100.1";"445.58648681640625"
"1580985240000";"server2";"446.2848536173503"
"1580985240000";"server1";"446.25459543863934"
"1580985240000";"10.1.100.10";"446.56190999348956"
"1580985240000";"10.1.100.1";"446.7175699869792"
```

In order to minimally impact existing queries, we can update timeGroup to select time column. Users will still have to update when not using macros. Will also need to make sure we can always select into time objects from our all of our sql datasources to have a proper schema.

So the SQL query becomes the following (having sql create time):

```sql
SELECT
  FROM_UNIXTIME(UNIX_TIMESTAMP(createdAt) DIV 60 * 60) as time_sec,
  avg(value) as value,
  hostname as metric
FROM 
  grafana_metric
WHERE
  createdAt BETWEEN FROM_UNIXTIME(1580908377) AND FROM_UNIXTIME(1580985308) AND
  measurement = 'logins.count' AND
  hostname IN('10.1.100.1','10.1.100.10','server1','server2')
GROUP BY 1, 3
ORDER BY 1
```

## Distinguishing Aspects Formats

- Both:
  - Have Time column
  - Have one or more Numeric Field
- Long
  - Will have String Fields (in Result)
  - Time will be repeated (Not detectable in schema, may be sparse)
  - Have no Labels on the Dataframe's Fields.
- Wide
  - Will Not have string Fields
  - Time (should) not be repeated
  - Optional labels

## How Long Format Time Series format is interpreted

Reference from azure data explorer datasource: https://github.com/grafana/azure-data-explorer-datasource#time-series-queries

- Has one Time column.
- Number type columns are "Metrics". The name of the column is the name of the metric.
- String type columns are Key/Value pair identifiers (a.k.a labels, tags). The column name is the identifer key, the value in each row is identifier Value.
- Each unique time series in the response is the unique combination of metric and Key/Value pair identifiers.

## How Wide Format Time Series format is interpret

- Has one Time column.
- Has one or more Number value columns. The Unique name of the column is the Metric. (**Edit:** Thinking about this more the name probably doesn't need to be unique since columns/fields are positional (have an index) which can be grouped on.)
- If the column has a Labels property, they become Key/Value pair identifiers. And the Name does not have to be unique if Name+Labels is unique.

## Misc notes and Reference

- Alerting and GEL will need same interpretation ability.
- There are some other types of time series to support. Those generally involve nested arrays - dataframe does have those yet.
- Wide vs Long is determined by format of query result, not format of data in database.
- There may be cases where there is a second time column, and maybe that should be treated as a graph-able value.
- We should not force these schemas to always be time series. Assumption is the context is one of time series.
- A reference on wide vs long: https://docs.timescale.com/latest/introduction/data-model
- Distinction might make table transforms more clear?
- Columns are time because backend Arrow type is time (which would also map to time.Time in the Go sdk).
- In order to break less queries in existing queries. We probably can at least do some stuff based column names of "time". Best to move away from this - maybe a legacy checkbox with deprecation?

**Comment by kylebrandt** [comment]
> There are some other types of time series to support. Those generally involve nested arrays - dataframe does have those yet.

One that doesn't involve nested arrays. The Factor Columns (aka dataframe Fields) are mixed with a schema where additional column *names* are the timestamps:

![image](https://user-images.githubusercontent.com/1692624/76521857-000eb280-643c-11ea-9784-cb2d799bb601.png)

**Comment by torkelo** [comment]
Moving milestone on this to 7.1 if that is ok?

**Comment by marefr** [comment]
@kylebrandt moving this to 7.2 for now since I'm not sure about the state.

**Comment by torkelo** [comment]
Not sure what is involved in solving this, and how much work it will be

**Comment by marefr** [comment]
@kylebrandt is this still outstanding? Any short term plans for solving this?

**Comment by torkelo** [comment]
dont we support this already?

**Comment by Rotario** [comment]
Hi All,
Thanks for your great work!
I'm trying to plot long format data on a single timeseries, and there seems to be no way to do it - is this still under development?
It seems like a critical use-case - I don't know if I've missed something.
I just can't product multiple time series from a DB with the following schema:
```
|time|foreign_id|value|
|xxxx|1              |2       |
|xxxx|2              |3      |
```
I've also asked the Q in discourse here https://community.grafana.com/t/plot-long-format-data/49901

I'd appreciate some confirmation that this is or isn't possible please? The docs are pretty confusing, and I'd like to contribute to them if that would help clear it up.
For example this issue is referenced here https://grafana.com/docs/grafana/latest/developers/plugins/data-frames/
specifically stating
"Grafana can detect and convert data frames in long format into wide format." but doesn't explain or got into further detail if this is a configuration, or happens automatically.
I'd appreciate some guidance please?
Thanks,
R

**Comment by svet-b** [comment]
Wondering if this still something that might be implemented in the near future? (I assume the 8.1.0 milestone is aspirational :) )

We have a dataset that looks exactly like the table in the issue description - time field, value field, and metric field. This is the result of a transformation that merges two other datasets. At present there doesn't appear to be a way to graph this as time series data (where the series are grouped by the "metric" field).

In theory the "[Prepare time series](https://grafana.com/docs/grafana/latest/panels/transformations/types-options/#prepare-time-series)" transformation would enable the outcome we need, via the "Wide time series" option, which would "transform the time series data frame from the long to the wide format". But selecting this option doesn't appear to do anything to the data.

I assume that this issue is essentially the blocker here? Or could there be another way to get the outcome we need, which isn't dependent on the implementation in question here?

**Comment by svet-b** [comment]
FWIW, [this PR](https://github.com/grafana/grafana/pull/38670) (which landed today in v8.2.0-beta1) enables the use case that I was after, via the "Multi-frame time series" option of the "Prepare time series" transform.

**Comment by kylebrandt** [comment]
Closing in favor of newer issue https://github.com/grafana/grafana/issues/57543 , which is very much the same thing, but in the context of the [dataplane contact](https://github.com/grafana/grafana-plugin-sdk-go/tree/main/data/contract_docs)
